src/libraries/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -12,24 +12,52 @@ public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> fir
         {
             if (first == null)
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.first);
+                ThrowHelper.ThrowArgumentNullException(nameof(first));
             }
 
             if (second == null)
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.second);
+                ThrowHelper.ThrowArgumentNullException(nameof(second));
             }
 
             return first is ConcatIterator<TSource> firstConcat
                 ? firstConcat.Concat(second)
-                : new Concat2Iterator<TSource>(first, second);
+                : new BaseConcatIterator<TSource>(first, second, rest: null);
         }
 
         /// <summary>
-        /// Represents the concatenation of two <see cref="IEnumerable{TSource}"/>.
+        /// Concatenates two or more sequences.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of the input sequences.</typeparam>
+        /// <param name="first">The first sequence to concatenate.</param>
+        /// <param name="second">The sequence to concatenate to the first sequence.</param>
+        /// <param name="rest">Specifies any subsequent sequences to concatenate.</param>
+        /// <returns>An <see cref="IEnumerable{T}"/> that contains the concatenated elements of the two input sequences.</returns>
+        public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second, params IEnumerable<TSource>[] rest)
```

Comment:
The whole purpose of this API is to allow someone to write:
```C#
a.Concat(b, c, d);
```
rather than:
```C#
a.Concat(b).Concat(c).Concat(d);
```
?  That doesn't seem worthwhile, in my opinion.

Reply:
I think there are arguments to be made about terseness and discoverability. It also allows concatenating arbitrarily many enumerables without needing to foreach over the inputs.

Reply:
>  It also allows concatenating arbitrarily many enumerables without needing to foreach over the inputs

That's what Flatten does.  This would only be used if you had two individual enumerables and an array of enumerables. That can't be common.

Reply:
It's slightly different than Flatten in that the number of flattened segments can be known ahead of enumeration, plus the optimizations applied by Concat's `Iterator<T>` implementation. I agree about the signature being awkward for this type of use. 

I still think `a.Concat(b,c,d)` is an improvement over `a.Concat(b).Concat(c).Concat(d)`.

Reply:
> It's slightly different than Flatten in that the number of flattened segments can be known ahead of enumeration

It can with Flatten as well, with TryGetNonEnumeratedCount.  That will benefit more than just arrays as well.

> I agree about the signature being awkward for this type of use.

More than awkward, I don't see it being used at all for that kind of use.  That would mean I have an `IEnumerable<T>[] source` and I write code like:
```C#
return source.Length switch
{
    0 => Enumerable.Empty<T>(),
    1 => source[0],
    2 => source[0].Concat(source[1]),
    _ => source[0].Concat(source[1], source[2..]),
}
```
which I don't imagine anyone will do.

That means the only reason this is being added is to be able to write:
```C#
a.Concat(b, c, d);
```
instead of:
```C#
a.Concat(b).Concat(c).Concat(d);
```
which, to me at least, is not worth it.  I understand it was already reviewed by others and folks agreed to add it, so it's up to you, but from a value perspective, I don't see this being worth a lot of code.

And if we do decide to keep it, FWIW I disagree with the separation of the second argument from the rest of the params.  Yes, it avoids someone writing the nop `e.Concat()`, but it also prevents someone from writing `a.Concat(array)` and instead forces them to slice that array to separate out the car and the cdr.  If we're going through the effort now and in the future to have this overload, I'd say the utility of supporting the array is more beneficial than preventing someone from writing `e.Concat()`.  There are many other mostly useless things folks can write with LINQ already that we don't prevent, e.g. `e.Select(i => i)`, `e.Where(i => true)`, etc.  To me, that's a job for an analyzer if we want to tell people that are doing things that aren't actually useful, ala https://github.com/dotnet/runtime/issues/33770.

Reply:
> That means the only reason this is being added is to be able to write:

There's also the theoretical performance improvement of avoiding a linked-list representation of the iterator cf. https://github.com/dotnet/runtime/pull/61230#discussion_r743579004

> FWIW I disagree with the separation of the second argument from the rest of the params.

Ultimately it's tradeoff between allowing `a.Concat()` and making `a.Concat(array)` possible. This wouldn't prevent slicing in your example though, you'd still have to write `source[0].Concat(source[1..])`

Reply:
> Ultimately it's tradeoff between allowing a.Concat() and making a.Concat(array) possible

Right. Why is the former deemed more negatively impactful than the latter is beneficial?

> This wouldn't prevent slicing in your example though

It would if I had a single enumerable and wanted to concat an array I get from somewhere. The current signature requires me to get _two_ single enumerables from somewhere plus that array.

Do we have concrete use examples for these APIs? Examples from around GitHub, etc.?  I'm curious what we used to drive the addition of these methods.

Reply:
I don't agree with the tradeoff.  But if all aspects of it were appropriately discussed and the overall opinion was this was the best tradeoff, ok.